File: environment/src/test/java/jetbrains/exodus/log/LogTestsBase.java
/**
 * Copyright 2010 - 2022 JetBrains s.r.o.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.log;

import jetbrains.exodus.ArrayByteIterable;
import jetbrains.exodus.TestUtil;
import jetbrains.exodus.core.dataStructures.Pair;
import jetbrains.exodus.io.*;
import jetbrains.exodus.util.IOUtil;
import org.junit.After;
import org.junit.Before;

import java.io.File;
import java.io.IOException;

class LogTestsBase {

    protected volatile Log log;
    private File logDirectory = null;
    protected DataReader reader;
    protected DataWriter writer;

    @Before
    public void setUp() throws IOException {
        SharedOpenFilesCache.setSize(16);
        final File testsDirectory = getLogDirectory();
        if (testsDirectory.exists()) {
            IOUtil.deleteRecursively(testsDirectory);
        } else if (!testsDirectory.mkdir()) {
            throw new IOException("Failed to create directory for tests.");
        }

        synchronized (this) {
            final Pair<DataReader, DataWriter> logRW = createLogRW();
            reader = logRW.getFirst();
            writer = logRW.getSecond();
        }
    }

    @After
    public void tearDown() {
        closeLog();
        final File testsDirectory = getLogDirectory();
        IOUtil.deleteRecursively(testsDirectory);
        IOUtil.deleteFile(testsDirectory);
        logDirectory = null;
    }

    protected Pair<DataReader, DataWriter> createLogRW() {
        FileDataReader reader = new FileDataReader(logDirectory);
        return new Pair<>(reader, new FileDataWriter(reader));
    }

    void initLog(final long fileSize, final int cachePageSize) {
        initLog(new LogConfig().setFileSize(fileSize).setCachePageSize(cachePageSize));
    }

    void initLog(final LogConfig config) {
        if (log == null) {
            synchronized (this) {
                if (log == null) {
                    log = new Log(config.setReaderWriter(reader, writer));
                }
            }
        }
    }

    protected Log getLog() {
        if (log == null) {
            synchronized (this) {
                if (log == null) {
                    log = new Log(LogConfig.create(reader, writer));
                }
            }
        }
        return log;
    }

    void closeLog() {
        if (log != null) {
            log.close();
            log = null;
        }
    }

    File getLogDirectory() {
        if (logDirectory == null) {
            logDirectory = TestUtil.createTempDir();
        }
        return logDirectory;
    }

    static TestLoggable createOneKbLoggable() {
        return new TestLoggable((byte) 126, new ArrayByteIterable(new byte[1024], 1024), Loggable.NO_STRUCTURE_ID);
    }

}

File: environment/src/test/kotlin/jetbrains/exodus/log/LogTestsBase.kt
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.log

import jetbrains.exodus.*
import jetbrains.exodus.core.dataStructures.Pair
import jetbrains.exodus.env.EnvironmentImpl
import jetbrains.exodus.io.*
import jetbrains.exodus.io.SharedOpenFilesCache.Companion.setSize
import jetbrains.exodus.log.LogConfig.Companion.create
import jetbrains.exodus.util.IOUtil.deleteFile
import jetbrains.exodus.util.IOUtil.deleteRecursively
import org.junit.After
import org.junit.Before
import java.io.*

open class LogTestsBase {
    @Volatile
    private var _log: Log? = null
    protected val log: Log
        get() {
            if (_log == null) {
                synchronized(this) {
                    if (_log == null) {
                        Log.invalidateSharedCache()
                        _log = Log(create(reader!!, writer!!), EnvironmentImpl.CURRENT_FORMAT_VERSION)
                    }
                }
            }

        return _log!!
    }

    private var _logDirectory: File? = null
    protected val logDirectory: File
        get() {
            if (_logDirectory == null) {
                _logDirectory = TestUtil.createTempDir()
            }
            return _logDirectory!!
        }

    protected var reader: DataReader? = null
    protected var writer: DataWriter? = null

    protected fun clearLog() {
        _log = null
    }

    @Before
    @Throws(IOException::class)
    fun setUp() {
        setSize(16)
        val testsDirectory = _logDirectory

        if (testsDirectory != null) {
            if (testsDirectory.exists()) {
                deleteRecursively(testsDirectory)
            } else if (!testsDirectory.mkdir()) {
                throw IOException("Failed to create directory for tests.")
            }
        }

        synchronized(this) {
            val logRW = createLogRW()
            reader = logRW.getFirst()
            writer = logRW.getSecond()
        }
    }

    @After
    fun tearDown() {
        closeLog()
        val testsDirectory = _logDirectory
        if (testsDirectory != null) {
            deleteRecursively(testsDirectory)
            deleteFile(testsDirectory)
            _logDirectory = null
        }
    }

    protected open fun createLogRW(): Pair<DataReader, DataWriter> {
        val reader = FileDataReader(logDirectory)
        return Pair(reader, AsyncFileDataWriter(reader))
    }

    fun initLog(fileSize: Long, cachePageSize: Int) {
        initLog(LogConfig().setFileSize(fileSize).setCachePageSize(cachePageSize))
    }

    private fun initLog(config: LogConfig) {
        if (_log == null) {
            synchronized(this) {
                if (_log == null) {
                    Log.invalidateSharedCache()
                    _log = Log(config.setReaderWriter(reader!!, writer!!), EnvironmentImpl.CURRENT_FORMAT_VERSION)
                }
            }
        }
    }

    fun closeLog() {
        if (_log != null) {
            _log!!.close()
            _log = null
        }
    }

    companion object {
        fun createOneKbLoggable(): TestLoggable {
            return TestLoggable(126.toByte(), ArrayByteIterable(ByteArray(1024), 1024), Loggable.NO_STRUCTURE_ID)
        }
    }
}
